Source/WebCore/page/text-extraction/TextExtraction.cpp

```diff
@@ -0,0 +1,189 @@
+    RefPtr<Text> lastTextNode;
+    StringBuilder textForLastTextNode;
+    for (TextIterator iterator { makeRangeSelectingNodeContents(container) }; !iterator.atEnd(); iterator.advance()) {
+        if (iterator.text().isEmpty())
+            continue;
+
+        RefPtr textNode = dynamicDowncast<Text>(iterator.node());
+        if (!textNode)
+            continue;
+
+        if (!lastTextNode)
+            lastTextNode = textNode;
+
+        if (lastTextNode == textNode) {
+            textForLastTextNode.append(iterator.text());
+            continue;
+        }
+
+        if (auto text = textForLastTextNode.toString().trim(isASCIIWhitespace<UChar>); !text.isEmpty())
+            nodeToTextMap.add(*lastTextNode, textForLastTextNode.toString());
+
+        textForLastTextNode.clear();
+        textForLastTextNode.append(iterator.text());
+        lastTextNode = textNode;
+    }
+
+    if (auto text = textForLastTextNode.toString().trim(isASCIIWhitespace<UChar>); lastTextNode && !text.isEmpty())
+        nodeToTextMap.add(*lastTextNode, textForLastTextNode.toString());
+
+    return nodeToTextMap;
+}
+
+static bool shouldIncludeChildren(const ItemData& data)
+{
+    return WTF::switchOn(data,
+        [&](const TextItemData&) { return false; },
+        [&](const ScrollableItemData&) { return true; },
+        [&](const EditableItemData&) { return true; },
+        [&](const ImageItemData&) { return false; },
+        [&](const InteractiveItemData&) { return true; },
+        [&](ContainerType) { return true; }
+    );
+}
+
+static inline FloatRect rootViewBounds(Node& node)
+{
+    auto view = node.document().protectedView();
+    if (UNLIKELY(!view))
+        return { };
+
+    if (!node.renderer())
+        return { };
+
+    return view->contentsToRootView(node.renderer()->absoluteBoundingBoxRect());
+}
+
+static inline std::optional<ItemData> extractItemData(Node& node, const HashMap<Ref<Text>, String>& extractedText)
+{
+    if (!node.renderer())
+        return std::nullopt;
+
+    if (RefPtr textNode = dynamicDowncast<Text>(node)) {
+        if (auto iterator = extractedText.find(*textNode); iterator != extractedText.end())
+            return { { TextItemData { iterator->value } } };
+        return std::nullopt;
+    }
+
+    RefPtr element = dynamicDowncast<Element>(node);
+    if (!element)
+        return std::nullopt;
+
+    if (!element->isInUserAgentShadowTree() && element->isRootEditableElement())
+        return { { EditableItemData { element == element->document().activeElement() } } };
+
+    if (RefPtr image = dynamicDowncast<HTMLImageElement>(element))
+        return { { ImageItemData { image->src().lastPathComponent().toString(), image->altText() } } };
+
+    if (RefPtr button = dynamicDowncast<HTMLButtonElement>(element))
+        return { { InteractiveItemData { !button->isDisabledFormControl() } } };
+
+    if (RefPtr input = dynamicDowncast<HTMLInputElement>(element)) {
+        if (input->isSubmitButton())
```

Comment:
Should we use `input->isTextButton()` here?

`<input type=button>` won't be included for `isSubmitButton()`, and I don't see a good reason to exclude it, given that `HTMLButtonElement` is included. Unless, you also wanted to tie the inclusion `HTMLButtonElement` to form submission?

Reply:
Great catch — it should definitely be `isTextButton` instead of `isSubmitButton`.